### ISSUE TITLE
[Feature Request] Smart Playlist's Code Box (Advanced Settings) 

### ISSUE THREAD

**Issue by 10bn** [ampache/ampache#2543]
Dear Devs,

let me start with a picture. 

<img width="1125" alt="Screenshot 2020-09-19 at 14 13 21" src="https://user-images.githubusercontent.com/46263124/93667339-6dec0480-fa85-11ea-85e2-4e96dc3b9950.png">

Could you consider adding as an advanced feature a code box below the smart playlist builder. 
The box is supposed to represent the code behind the set rules and lets the user add custom code (playlist rules)?

**Comment by lachlan-00** [comment]
you mean it would load the sql to allow editing the query?

**Comment by jaddel** [comment]
yes, exactly.

**Comment by 4phun** [comment]
I second this, great idea.

**Comment by AshotN** [comment]
Makes you think we should allow advanced rules to be exported and imported.

**Comment by kuzi-moto** [comment]
This would be a neat idea. I'm assuming it's easy enough to check the input though to make sure someone isn't able to drop all tables, or pull the user information or something?

**Comment by 4phun** [comment]
@kuzi-moto You're right, this can be a huge security breach. Maybe it should be shown only for administrators.

**Comment by lachlan-00** [comment]
i can get why you'd want to do this and as an admin i would just end up doing it in sql manually. (i used to run my own sql files to make my playlists) but it would be better to not give anyone a sql query in a text box from the UI.

It would be good to know what you'd actually use it fore besides generating new rule types.

**Comment by jaddel** [comment]
It would simply be very convenient to be able to do this in the web ui and build advanced smart playlists without editing tables in "sql". 
I'm not a developer but somethings are just a lot simpler to do by a few lines of code. 

For some people I could image something like blockly would be perfect to build fancy smart playlist algorithms.

**Comment by lachlan-00** [comment]
I'm saying you use sql when the rules don't exist but if the rules do exist i don't really see that adding a second way to do the same thing is a benefit.

**Comment by AshotN** [comment]
> This would be a neat idea. I'm assuming it's easy enough to check the input though to make sure someone isn't able to drop all tables, or pull the user information or something?

I don't think it would be to bad. Allow admins to create custom search templates. Block commands such as DROP or INSERT.

**Comment by jaddel** [comment]
> I'm saying you use sql when the rules don't exist but if the rules do exist i don't really see that adding a second way to do the same thing is a benefit.

It depends a bit on the complexity of what you want to do. 

When I asked for the code box i was thinking of the advanced casual user, who has just enough knowledge to maybe setup Ampache in a docker. I believe doing this in sql would be a barrier for many users / admins and some are just users who don't do any kind of admin stuff. 

Surely its not a necessity.

**Comment by usox** [comment]
We definitely need some kind of abstraction here - not just for security reasons. If users can edit sql in there, it would make database migrations during updates hard/impossible without breaking a users playlist filter. Also, this could have a huge performance impact, depending on what the user builds (in huge installations).
TBH I've to deal with something like this in my current job and I strongly recommend to be as restrictive a possible - this is some kind of `pandora's box`. Once opened, we loose control over the way, data gets queried and this can end up bad.

**Comment by AshotN** [comment]
To be fair though, Ampache is mostly for personal self-hosted applications. So if a user decides to enable the feature they should understand the risks. But I think abstraction could be better from a usability standpoint, so people don't have to know SQL.

**Comment by kuzi-moto** [comment]
At the very least, the ability to create custom rules can be put in settings so that by default only Admins can create them.

**Comment by AshotN** [comment]
I really like the idea, but I honestly think it's going to take a lot of deliberation on how to do it properly. If someone is willing to draft a concept. That would help.

**Comment by skeagy** [comment]
I just made a feature request that is somewhat related... nested boolean rules in smart playlist creation:
https://github.com/ampache/ampache/issues/2995

If you take that, and a new rule type "match regex", you have a pretty rich flexibility.

At that point, the only reason I can see using raw sql would be if you have custom modded the DB with new tables or fields and importing data from other sources... which would be pretty cool. "Find all songs with singer who died before 30 years old." At that point, I'd just build the playlist in a script outside of Ampache and import the playlist.

That seems cleaner than dealing with the security issues of filtering raw SQL, and dealing with ways people try to bypass the filtering out of DROP, INSERT or other nefarious things like redefining tables, changing charsets, etc. People who try to evade security mechanisms on average seem to be more clever than people who try to block them.
